[ChongWeiJie29] iP

ChongWeiJie29 opened the pull request:
# ChampionSOS

 > The statement below is false. 
The statement above is true.
Learn more about [Liar's Paradox](https://en.wikipedia.org/wiki/Liar_paradox)!

ChampionSOS frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/ChongWeiJie29/ip/releases/download/A-Jar/duke.jar).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ChongWeiJie29 committed:
Complete Level 0: Rename, Greet, Exit

Rename chatbot.
Add in feature for bot to greet and exit.

ChongWeiJie29 committed:
Complete Level 1: Echo

Make bot echo command by user.
Exits only when user says bye.

ChongWeiJie29 committed:
Complete Level 2: Add, List

Add in ability to add text into an array.
Add in ability to return array of text added.

ChongWeiJie29 committed:
Complete Level 3: Mark as Done

Add ability to toggle tasks between done
and undone.

ChongWeiJie29 committed:
Complete Level 4: ToDos, Events, Deadlines

Add in specific tasks as categories, with
respective details such as timing of event
and 'due by' deadline

ChongWeiJie29 committed:
Complete A-TextUiTesting: Automated Text UI Testing

Added in files for semi-automated testing of Duke.

ChongWeiJie29 committed:
Complete Level 5: Handle Errors

Handle most incorrect inputs by users, such as empty task, or
improper formatting.

ChongWeiJie29 committed:
Supplement Level 5: Handle Errors

Added in a few more scenarios of incorrect inputs from users
such as empty 'from', 'to' and 'by'.

ChongWeiJie29 committed:
Complete Level 6: Delete

Add feature to delete task from list.

ChongWeiJie29 committed:
Additional modification for Level 6: Delete

Added in error handling for deletion.
Modified some error handling for marking and
unmarking for previous Level 5.

ChongWeiJie29 committed:
Close scanner object at the end

ChongWeiJie29 committed:
Complete Level 7: Save

Save task list in disk, and load
data into bot when starting.

ChongWeiJie29 committed:
Merge branch 'branch-Level-7'

ChongWeiJie29 committed:
Complete Level 8: Dates and Times

Change Duke bot to now understand date inputs
as a LocalDate Object rather than just a string.
Also, prints these details in a different format.

ChongWeiJie29 committed:
Merge branch 'branch-Level-8'

ChongWeiJie29 committed:
Complete A-OOP: Use More OOP

Refactor code to have Ui, Storage,
Parser and TaskList classes.

ChongWeiJie29 committed:
Merge branch 'branch-A-OOP'

ChongWeiJie29 committed:
Complete A-Packages: Organize into packages

Organise existing classes into suitable packages.

ChongWeiJie29 committed:
Merge branch 'branch-A-Packages'

ChongWeiJie29 committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

ChongWeiJie29 committed:
Complete A-Gradle: Use Gradle

Automate build task of project.

ChongWeiJie29 committed:
Merge branch 'branch-A-Gradle'

ChongWeiJie29 committed:
Complete A-JUnit: Add JUnit Tests

Add JUnit tests to test code.

ChongWeiJie29 committed:
Merge branch 'branch-A-JUnit'

ChongWeiJie29 committed:
Complete A-Jar: Create a JAR File

Package app as a jar file.

ChongWeiJie29 committed:
Merge branch 'branch-A-Jar'

ChongWeiJie29 committed:
Complete A-JavaDoc: JavaDoc

Add JavaDoc comments to methods

ChongWeiJie29 committed:
Merge branch 'branch-A-JavaDoc'

ChongWeiJie29 committed:
Complete A-CodingStandard: Follow the Coding Standard

Tweak code to comply with Java coding standard.

ChongWeiJie29 committed:
Merge branch 'branch-A-CodingStandard'

Merge A-CodingStandard into master

Merge branch into master after fixing merge conflicts.

ChongWeiJie29 committed:
Change gradle build file

Change build.gradle file to match the change
in package name.

ChongWeiJie29 committed:
Complete Level 9: Find

Add in feature to find tasks based on keyword.

ChongWeiJie29 committed:
Merge branch 'branch-Level-9'

Merge Level 9: Find

Merge Level-9 into master branch after
fixing merge conflicts

ChongWeiJie29 committed:
Add in JavaDoc comments for classes

Add in JavaDoc comments for classes, which
were missed out previously.

ChongWeiJie29 committed:
Complete Level 10: GUI

Add a GUI to the chatbot using JavaFX.

ChongWeiJie29 committed:
Merge branch 'branch-GUI'

Jweng88 commented on src/main/java/dukepackage/Duke.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class Duke {
+
+    private Storage storage;
```
Perhaps single line Javadoc comments can be added for the class members?

Jweng88 commented on src/main/java/toolpackage/Parser.java:
```diff
@@ -0,0 +1,71 @@
+            task = parsedCommand[1];
+        }
+
+        if (command.equals("bye")) {
```
Perhaps consider using switch statements?

Jweng88 commented on src/main/java/toolpackage/TaskList.java:
The naming of the methods are all of the course standard. I like how the code is neat and tidy with no unused parameters in method signatures or multiple statements in the same line. Good job! 👍

Jweng88 commented on src/test/java/toolpackage/StorageTest.java:
I like how you have thoroughly tested the methods in the file. Looks good to me! 👍

Jweng88 commented on src/main/java/taskpackage/Task.java:
I like the clarity of the class and methods as they are short with little nesting. Good job! 👍

Jweng88 commented on src/main/java/dukepackage/DukeException.java:
I like the use of custom exceptions as a layer of abstraction to handle potential errors. Good job! 👍

Jweng88 commented on src/test/java/toolpackage/TaskListTest.java:
I like the detailed testing of different methods in the file with no unused code. Good job! 👍

Jweng88 commented on src/main/java/toolpackage/Ui.java:
```diff
@@ -0,0 +1,113 @@
+ */
+public class Ui {
+
+    private static final String TEXT_GREETING = "Hello! I'm ChampionSOS\nWhat can I do for you?";
```
I like the use of constants for greater clarity in the code. Looks good to me! 👍

Jweng88 reviewed (commented):
Looks good to me! Good job with the code quality!

nananakx-x commented on src/main/java/toolpackage/DialogBox.java:
```diff
@@ -0,0 +1,85 @@
+package toolpackage;
```
I like that you created a separate package for gui and dialogbox, good job!

nananakx-x commented on src/main/java/toolpackage/Parser.java:
```diff
@@ -0,0 +1,75 @@
+            task = parsedCommand[1];
+        }
+
+        if (command.equals("bye")) {
```
Agree with the other review that perhaps switch statements would improve readability

nananakx-x commented on src/test/java/toolpackage/TaskListTest.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Test
+    public void removeItem_incorrectIndex_showErrorMsg() {
```
Tests took many different cases and their expected output into consideration, I think this is really well done :)

nananakx-x commented on src/main/java/toolpackage/Ui.java:
```diff
@@ -0,0 +1,120 @@
+package toolpackage;
+
+import taskpackage.Task;
+
```
Use of line spacings to separate import statements improves readability

nananakx-x commented on src/main/java/taskpackage/Deadlines.java:
```diff
@@ -0,0 +1,76 @@
+            this.by = LocalDate.parse(details.substring(3));
+        } catch (DateTimeParseException e) {
+            byte[] emojiByteCode = new byte[]{(byte)0xF0, (byte)0x9F, (byte)0x98, (byte)0xB1};
+            String emoji = new String(emojiByteCode, Charset.forName("UTF-8"));
```
Interesting method using byte, I learnt a lot from your code!

nananakx-x reviewed (commented):
Overall LGTM! Coding standard was strictly adhered to and javadocs was done for almost every method, good job on that!! Perhaps it would be better to delete methods which are no longer in use to make it more concise, but other than that, I find it really neat and tidy :")

ChongWeiJie29 committed:
Add to Level 10: GUI

Change profile picture of bot and user.

ChongWeiJie29 committed:
Complete A-CodeQuality: Improve Code Quality

Currently, there are certain coding practices that are not corrected.
This reduces the readability of the code for other programmers.

To correct this, the code has been refactored to enhance organisation
so that there is less code repetitions and a more coherent flow,
making it easier for other programmers to trace the code.

ChongWeiJie29 committed:
Complete A-Assertions: Use Assertions

There is currently insufficient checking of inputs and
variables to ensure that assumptions hold. This would be
problematic if the assumptions fail to hold, as it results
in incorrect running of the chatbot.

To rectify this, assert statements should be added to test
for such assumptions. This will signal to programmers in the event the
assumption fails to hold, enabling programmers to rectify the error.

ChongWeiJie29 committed:
Merge pull request #1 from ChongWeiJie29/branch-A-Assertions

Complete A-Assertions: Use Assertions

ChongWeiJie29 committed:
Remove irrelevant imports

ChongWeiJie29 committed:
Merge branch 'master' into branch-A-CodeQuality

ChongWeiJie29 committed:
Merge pull request #2 from ChongWeiJie29/branch-A-CodeQuality

Branch a code quality

ChongWeiJie29 committed:
Add an extention B-Reminders

There is currently no way to track whether the due
date of the deadline or event is getting close. Users
may have to manually search up the entire list to
see which is due soon.

This feature will help to filter out tasks
due soon, and remind users of these tasks and their
deadline upon start up and also upon the use of the
'remind' command.

ChongWeiJie29 committed:
Merge branch 'branch-B-Reminders'

ChongWeiJie29 committed:
Add in final changes.

ChongWeiJie29 committed:
Merge branch 'b-finalChanges'

ChongWeiJie29 committed:
Fix bug found in test case.

ChongWeiJie29 committed:
Merge branch 'b-finalChanges'

ChongWeiJie29 committed:
Add Ui.png file into docs

ChongWeiJie29 committed:
Add in user guide content in docs

ChongWeiJie29 committed:
Merge branch 'branch-User-Guide'

ChongWeiJie29 committed:
change slight error in User Guide

ChongWeiJie29 committed:
Merge branch 'branch-User-Guide'
